internal/resource/buffer.go

```diff
@@ -0,0 +1,110 @@
+package resource
+
+import (
+	"errors"
+	"io"
+)
+
+var (
+	ErrOutOfMemory    = errors.New("unable to grow buffer")
+	ErrOffsetTooLarge = errors.New("cannot seek past end of buffer")
+	ErrUnknownWhence  = errors.New("unknown whence")
+)
+
+type buffer struct {
+	buf    []byte
+	cursor int64
+	end    int64
+	cap    int64
```

Comment:
I'd then need to use a mixture of `copy` and `append` when copying in data, since the slice itself would need to be tweaked to cover the new data. This seemed cleaner to me.